### PR TITLE
Harden the code against protobuf exceptions on bad messages from peers

### PR DESCRIPTION
**Pull request by cculianu**
protobuf ParseFromString may raise exceptions which this code wasn't catching.

It's entirely possible for peers to send a malformed message intentionally... we attempt to deal with it and blame when we can.

### CODE REVIEW

**Comment by cculianu** [comment]
Ok, well.. I reviewed this and tested it. Merging.

File: plugins/shuffle/round.py [hunk 1]
```diff
@@ -227,41 +239,45 @@ def process_shuffling(self):
...
-            if self.inbox[phase].get(sender):
-                self.messages.packets.ParseFromString(self.inbox[phase][sender])
-                try:
-                    for packet in self.messages.packets.packet:
-                        packet.packet.message.str = self.crypto.decrypt(packet.packet.message.str)
-                    if not self.different_ciphertexts():
-                        self.messages.add_str(self.encrypt_new_address())
+            msg = self.inbox[phase].get(sender)
+            if msg:
+                accused = self._parse_from_string_safe(msg, sender)
```
Comment (emergent-reasons) [inline plugins/shuffle/round.py:247]: result if the message parses correctly: `accused` = `None`
result if the message fails: `accused` == `sender`

This logic is hard to follow but I think it is correct in this context. The usage of the function is kind of inverted from the way it is used in the announcement method.

File: plugins/shuffle/round.py [hunk 2]
```diff
@@ -227,41 +239,45 @@ def process_shuffling(self):
         4. If player is last player he goes to the next phase ("Broadcasts Output") and broadcast the outputs.
         """
         phase = self.messages.phases[self.phase]
-        accused = None
+        accused, sender = None, None
         if self.me == self.last_player():
             sender = self.players[self.previous_player(player=self.last_player())]
```
Comment (emergent-reasons) [inline plugins/shuffle/round.py:244]: I know you didn't write this but this and the similar statement in the `else:` should be exactly equivalent. If not, it is surprising logic.

```python
if self.me == self.last_player():
    sender = self.players[self.previous_player(player=self.last_player())]
else:
    sender = self.players[self.previous_player()]
```

File: plugins/shuffle/round.py [hunk 3]
```diff
@@ -215,6 +217,16 @@ def process_announcement(self):
                     self.log_message("encrypt new address")
                     self.phase = 'BroadcastOutput'
 
+    def _parse_from_string_safe(self, msg, sender):
+        ''' Parses messages from 'string' msg (really bytes).
+        Returns sender if there is a decode error, otherwise returns None '''
+        try:
+            self.messages.packets.ParseFromString(msg)
```
Comment (emergent-reasons) [inline plugins/shuffle/round.py:224]: Is this implemented in protobuf/text_format.py ? Or somewhere else?
This will work to avoid errors. If the right message never shows up, I assume the caller of this chain will time out and give up.

File: plugins/shuffle/round.py [hunk 4]
```diff
@@ -199,8 +199,10 @@ def process_announcement(self):
             messages = self.inbox[phase]
             self.encryption_keys = dict()
             self.change_addresses = {}
-            for message in messages:
-                self.messages.packets.ParseFromString(messages[message])
+            for message, msg in messages.items():
+                bad = self._parse_from_string_safe(msg, 'bad')
+                if bad:
+                    continue
```
Comment (emergent-reasons) [inline plugins/shuffle/round.py:205]: We don't abort protocol if we receive a bad message?

Just an efficiency issue but since we don't remove the bad message, whatever parent loop is calling this will qualify every time as "full".

File: plugins/shuffle/round.py [hunk 5]
```diff
@@ -227,41 +239,45 @@ def process_shuffling(self):
...
+                        else:
+                            accused = sender
+                    except CryptoError:
                         accused = sender
-                except CryptoError:
-                    accused = sender
         if accused is not None:
             self.skipped_equivocation_check(accused)
-            self.log_message("wrong from " + str(accused))
+            self.log_message("bad shuffle message from " + str(accused))
```
Comment (emergent-reasons) [inline plugins/shuffle/round.py:280]: accused can only be `sender` with the above logic. Is that right?

File: plugins/shuffle/round.py (hunk 1 above)
Comment (emergent-reasons) [reply plugins/shuffle/round.py:247]: Nevermind. I see that this is the way it's really intended to be used. the `"bad"` name in the other method was a little confusing.

Comment (cculianu) [reply plugins/shuffle/round.py:247]: I agree it's hard to follow .. but I had to write a function that fit reasonably within the confines of this half-assed code.

:/

File: plugins/shuffle/round.py (hunk 2 above)
Comment (cculianu) [reply plugins/shuffle/round.py:244]: Yah I saw that too.. it looked on the surface like the else and the if are the same exact thing.   This is written in the most annoying way possible. :/

File: plugins/shuffle/round.py (hunk 3 above)
Comment (cculianu) [reply plugins/shuffle/round.py:224]: Yeah it's a protobuf function that raises on badly formatted messages.  he never bothered to ever check for it, hence the need for this.

Yeah ultimately either there will be an immediate abort-with-blame or in some cases timeout.

The code is not very resilient to bad/garbage messages.  Likely there are still extant holes.  Just trying to plug up what I can. Bad actors shouldn't be causing clients to have exception crash reporter windows pop up willy nilly.  That's just silly.

File: plugins/shuffle/round.py (hunk 4 above)
Comment (cculianu) [reply plugins/shuffle/round.py:205]: Not always.  Sometimes you know "who to blame" so I tried to blame where I could (since the message has an originating key).

In some cases an abort is what makes sense...

File: plugins/shuffle/round.py (hunk 5 above)
Comment (cculianu) [reply plugins/shuffle/round.py:280]: Yeah either that or None. :)
